src/engine/Engine.h

```diff
@@ -302,13 +267,22 @@ class Engine {
       // Intersect both lists.
       size_t i = 0;
       size_t j = 0;
-      while (a(i, jc1) < sent1) {
+      // while (a(i, jc1) < sent1) {
+      while (i < a.size() && j < b.size()) {
         while (a(i, jc1) < b(j, jc2)) {
           ++i;
+          if (i >= a.size()) {
+            goto finish;
+          }
         }
+
         while (b(j, jc2) < a(i, jc1)) {
           ++j;
+          if (j >= b.size()) {
+            goto finish;
+          }
         }
+
         while (a(i, jc1) == b(j, jc2)) {
           // In case of match, create cross-product
```

Comment:
If I am not mistaken, the inner while loop that handles the cross-product is always the same
for all join-related functions. With inlining etc. I should not cost performance, and if we can measure.

Reply:
The only problem with this is that it manipulates `i` and `j` and thus we would either have to pass them as `size_t*` or have it return e.g. a `std::pair<size_t, size_t>` and kind of awkwardly update `i` and `j` with that. Sadly one can't assign in a structured binding.

Reply:
Speaking of killing code duplication, I think we should try to merge the two galloping cases. I'll give it a try